# Hospital City Data Anaylsis

## 1. Data Understanding

### 1.2 Preparation

In [62]:
import pandas as pd
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

#### 1.2.1 Import the dataset

In [63]:
patient_data=pd.read_excel('Project Data.xlsx')

In [64]:
patient_data.head()

Patient ID   Age  Gender City of Residence Profession Insurance Provider  \
0        1482  65.0  Female        Birmingham    Retired                NaN   
1        1482  65.0  Female        Birmingham    Retired                NaN   
2        1482  65.0  Female        Birmingham    Retired                NaN   
3        1482  65.0  Female        Birmingham    Retired                NaN   
4        1482  65.0  Female        Birmingham    Retired                NaN   

  Family History Education Level Marital Status Visit Date     Department  \
0  Heart Disease          Master       Divorced 2024-02-22  Endocrinology   
1  Heart Disease          Master       Divorced 2024-01-21      Emergency   
2  Heart Disease          Master       Divorced 2024-06-13      Neurology   
3  Heart Disease          Master       Divorced 2024-03-24      Emergency   
4  Heart Disease          Master       Divorced 2024-03-19      Neurology   

   Consultation Duration  Satisfaction Level  Approximate Annual Income  \
0                    110                   2                    38839.0   
1                     76                   4                    52219.0   
2                     17                   5                    41491.0   
3                     15                   3                    34944.0   
4                     99                   1                    39248.0   

   Consultation Price  Insurance Coverage  
0          293.014042                 0.0  
1          287.208982                 0.0  
2          241.254507                 0.0  
3          221.177483                 0.0  
4          219.743794                 0.0

#### 1.2.2 Fixing the index 

In [65]:
patient_data=patient_data.set_index('Patient ID')


In [66]:
patient_data.head()

Age  Gender City of Residence Profession Insurance Provider  \
Patient ID                                                                 
1482        65.0  Female        Birmingham    Retired                NaN   
1482        65.0  Female        Birmingham    Retired                NaN   
1482        65.0  Female        Birmingham    Retired                NaN   
1482        65.0  Female        Birmingham    Retired                NaN   
1482        65.0  Female        Birmingham    Retired                NaN   

           Family History Education Level Marital Status Visit Date  \
Patient ID                                                            
1482        Heart Disease          Master       Divorced 2024-02-22   
1482        Heart Disease          Master       Divorced 2024-01-21   
1482        Heart Disease          Master       Divorced 2024-06-13   
1482        Heart Disease          Master       Divorced 2024-03-24   
1482        Heart Disease          Master       Divorced 2024-03-19   

               Department  Consultation Duration  Satisfaction Level  \
Patient ID                                                             
1482        Endocrinology                    110                   2   
1482            Emergency                     76                   4   
1482            Neurology                     17                   5   
1482            Emergency                     15                   3   
1482            Neurology                     99                   1   

            Approximate Annual Income  Consultation Price  Insurance Coverage  
Patient ID                                                                     
1482                          38839.0          293.014042                 0.0  
1482                          52219.0          287.208982                 0.0  
1482                          41491.0          241.254507                 0.0  
1482                          34944.0          221.177483                 0.0  
1482                          39248.0          219.743794                 0.0

### 1.3 Data Inspection

In [67]:
patient_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10008 entries, 1482 to 1012
Data columns (total 15 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   Age                        9952 non-null   float64       
 1   Gender                     10008 non-null  object        
 2   City of Residence          10008 non-null  object        
 3   Profession                 10008 non-null  object        
 4   Insurance Provider         8098 non-null   object        
 5   Family History             8015 non-null   object        
 6   Education Level            9979 non-null   object        
 7   Marital Status             10008 non-null  object        
 8   Visit Date                 10008 non-null  datetime64[ns]
 9   Department                 10008 non-null  object        
 10  Consultation Duration      10008 non-null  int64         
 11  Satisfaction Level         10008 non-null  int64         
 12  Approxi

We have **8 categorical variables**:  
`Gender`, `City of Residence`, `Profession`, `Insurance Provider`, `Family History`, `Education Level`, `Marital Status`, and `Department`.

We also have **6 numerical variables**:  
`Age`, `Consultation Duration`, `Satisfaction Level`, `Approximate Annual Income`, `Consultation Price`, and `Insurance Coverage`.

Additionally, there is **1 datetime variable**:  
`Visit Date`.


In [68]:
patient_data.describe(include='all') #temos de tratar do range da idade

Age Gender City of Residence Profession Insurance Provider  \
count   9952.000000  10008             10008      10008               8098   
unique          NaN      3                 8         10                  4   
top             NaN  Other        Birmingham    Retired         Provider D   
freq            NaN   3417              1473       3601               2165   
mean      50.635651    NaN               NaN        NaN                NaN   
min        0.000000    NaN               NaN        NaN                NaN   
25%       23.000000    NaN               NaN        NaN                NaN   
50%       52.000000    NaN               NaN        NaN                NaN   
75%       75.000000    NaN               NaN        NaN                NaN   
max      195.000000    NaN               NaN        NaN                NaN   
std       31.185608    NaN               NaN        NaN                NaN   

       Family History Education Level Marital Status  \
count            8015            9979          10008   
unique              4               8              4   
top     Heart Disease   Undergraduate       Divorced   
freq             2235            4160           2893   
mean              NaN             NaN            NaN   
min               NaN             NaN            NaN   
25%               NaN             NaN            NaN   
50%               NaN             NaN            NaN   
75%               NaN             NaN            NaN   
max               NaN             NaN            NaN   
std               NaN             NaN            NaN   

                           Visit Date  Department  Consultation Duration  \
count                           10008       10008           10008.000000   
unique                            NaN          13                    NaN   
top                               NaN  Psychiatry                    NaN   
freq                              NaN        1361                    NaN   
mean    2024-03-31 14:53:14.244604160         NaN              67.807654   
min               2024-01-01 00:00:00         NaN              15.000000   
25%               2024-02-15 00:00:00         NaN              41.000000   
50%               2024-03-31 00:00:00         NaN              68.000000   
75%               2024-05-16 00:00:00         NaN              94.000000   
max               2024-06-30 00:00:00         NaN             600.000000   
std                               NaN         NaN              32.447143   

        Satisfaction Level  Approximate Annual Income  Consultation Price  \
count         10008.000000               9.854000e+03        10008.000000   
unique                 NaN                        NaN                 NaN   
top                    NaN                        NaN                 NaN   
freq                   NaN                        NaN                 NaN   
mean              3.202338               4.340276e+04          187.263022   
min               1.000000               0.000000e+00           50.036762   
25%               2.000000               3.115775e+04          111.316562   
50%               3.000000               4.087650e+04          159.531115   
75%               5.000000               5.072175e+04          199.820999   
max               6.000000               1.197090e+07        39999.222018   
std               1.565507               2.681420e+05          862.865550   

        Insurance Coverage  
count          9958.000000  
unique                 NaN  
top                    NaN  
freq                   NaN  
mean            115.429366  
min               0.000000  
25%              59.692518  
50%             115.938998  
75%             165.437958  
max             421.887836  
std              79.337759

In [69]:
patient_data.columns

Index(['Age', 'Gender', 'City of Residence', 'Profession',
       'Insurance Provider', 'Family History', 'Education Level',
       'Marital Status', 'Visit Date', 'Department', 'Consultation Duration',
       'Satisfaction Level', 'Approximate Annual Income', 'Consultation Price',
       'Insurance Coverage'],
      dtype='object')

### 1.4 Optmizing datatypes to save space

In [70]:
patient_data['Visit Date'] = pd.to_datetime(patient_data['Visit Date'])

category_cols = [
    'Gender', 'City of Residence', 'Profession',
       'Insurance Provider', 'Family History', 'Education Level',
       'Marital Status', 'Department'
]
patient_data[category_cols] = patient_data[category_cols].astype("category")

patient_data['Consultation Duration'] = pd.to_timedelta(patient_data['Consultation Duration'])
patient_data['Satisfaction Level'] = patient_data['Satisfaction Level'].astype("int8")
patient_data['Approximate Annual Income'] = patient_data['Approximate Annual Income'].astype("float32")
patient_data['Consultation Price'] = patient_data['Consultation Price'].astype("float16")
patient_data['Insurance Coverage'] = patient_data['Insurance Coverage'].astype("float16")

patient_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10008 entries, 1482 to 1012
Data columns (total 15 columns):
 #   Column                     Non-Null Count  Dtype          
---  ------                     --------------  -----          
 0   Age                        9952 non-null   float64        
 1   Gender                     10008 non-null  category       
 2   City of Residence          10008 non-null  category       
 3   Profession                 10008 non-null  category       
 4   Insurance Provider         8098 non-null   category       
 5   Family History             8015 non-null   category       
 6   Education Level            9979 non-null   category       
 7   Marital Status             10008 non-null  category       
 8   Visit Date                 10008 non-null  datetime64[ns] 
 9   Department                 10008 non-null  category       
 10  Consultation Duration      10008 non-null  timedelta64[ns]
 11  Satisfaction Level         10008 non-null  int8          

### 1.5 Treating Missing Values and Outliers

#### 1.5.1 Duplicates

We first inspected the dataset shape and then removed duplicated rows:

In [71]:
patient_data.shape

(10008, 15)

In [72]:
patient_data = patient_data.drop_duplicates()
print("After removing duplicates:", patient_data.shape)

After removing duplicates: (10000, 15)


After removal, the dataset went from 10,008 to 10,000 rows, meaning 8 duplicates were found and removed.

#### 1.5.2 Missing Values

In [73]:
patient_data.isnull().sum()

Age                            56
Gender                          0
City of Residence               0
Profession                      0
Insurance Provider           1907
Family History               1990
Education Level                29
Marital Status                  0
Visit Date                      0
Department                      0
Consultation Duration           0
Satisfaction Level              0
Approximate Annual Income     154
Consultation Price              0
Insurance Coverage             49
dtype: int64

From the missing value check, only a few features contain null entries:

- **Age** — 56  
- **Insurance Provider** — 1907  
- **Family History** — 1990  
- **Education Level** — 29  
- **Approximate Annual Income** — 154  
- **Insurance Coverage** — 49  

All remaining variables have complete data.


#### Patient specific and visit level features

To handle missing values appropriately, we separated the features two lists. Some variables represent **stable patient characteristics**, while others correspond to **individual visit details**.  



In [ ]:
# Patient specific
pat_cols = [
    'Age','Insurance Provider', 'Family History', 'Education Level','Approximate Annual Income'
]

# Visit-level numeric fields
visit_num = ['Consultation Duration','Consultation Price','Insurance Coverage']

In [ ]:

# Fill within each Patient ID by propagating known values
patient_data[pat_cols] = patient_data.groupby('Patient ID')[pat_cols].transform(
    lambda x: x.ffill().bfill())

In [ ]:
for col in visit_num:
    patient_data[col] = patient_data.groupby('Department')[col].transform(
        lambda x: x.fillna(x.median()))

C:\Users\maria\AppData\Local\Temp\ipykernel_23108\1163201387.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  patient_data[col] = patient_data.groupby('Department')[col].transform(
C:\Users\maria\AppData\Local\Temp\ipykernel_23108\1163201387.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  patient_data[col] = patient_data.groupby('Department')[col].transform(
C:\Users\maria\AppData\Local\Temp\ipykernel_23108\1163201387.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True

In [78]:
patient_data.isnull().sum()

Age                             0
Gender                          0
City of Residence               0
Profession                      0
Insurance Provider           1857
Family History               1990
Education Level                 0
Marital Status                  0
Visit Date                      0
Department                      0
Consultation Duration           0
Satisfaction Level              0
Approximate Annual Income       0
Consultation Price              0
Insurance Coverage              0
dtype: int64

In [79]:
print(patient_data['Insurance Provider'].value_counts())
print('--------')
print(patient_data['Family History'].value_counts())

Insurance Provider
Provider D    2174
Provider A    2013
Provider B    2006
Provider C    1950
Name: count, dtype: int64
--------
Family History
Heart Disease    2234
Hypertension     2047
Diabetes         1866
Cancer           1863
Name: count, dtype: int64


In [80]:
# Add new categories first before filling
patient_data['Insurance Provider'] = patient_data['Insurance Provider'].cat.add_categories(['No Provider'])
patient_data['Family History'] = patient_data['Family History'].cat.add_categories(['No Disease'])

# Now fill with the new categories
patient_data['Insurance Provider'].fillna('No Provider', inplace=True)
patient_data['Family History'].fillna('No Disease', inplace=True)

# Verify no missing values remain
patient_data.isnull().sum()

C:\Users\maria\AppData\Local\Temp\ipykernel_23108\3422405668.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  patient_data['Insurance Provider'].fillna('No Provider', inplace=True)
C:\Users\maria\AppData\Local\Temp\ipykernel_23108\3422405668.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behav

Age                          0
Gender                       0
City of Residence            0
Profession                   0
Insurance Provider           0
Family History               0
Education Level              0
Marital Status               0
Visit Date                   0
Department                   0
Consultation Duration        0
Satisfaction Level           0
Approximate Annual Income    0
Consultation Price           0
Insurance Coverage           0
dtype: int64

In [81]:
print(patient_data['Insurance Provider'].value_counts())
print('--------')
print(patient_data['Family History'].value_counts())

Insurance Provider
Provider D     2174
Provider A     2013
Provider B     2006
Provider C     1950
No Provider    1857
Name: count, dtype: int64
--------
Family History
Heart Disease    2234
Hypertension     2047
No Disease       1990
Diabetes         1866
Cancer           1863
Name: count, dtype: int64
